_posts/2020-01-21-graphql-relay-windowed-pagination.markdown

```diff
@@ -0,0 +1,328 @@
+a prev/next navigation, which brings the user forward and back one page at a time. Whenever the current page
+changes, the URL should update accordingly. For a responsive implementation, we want to hide the page numbers, and
+only show the prev/next toggles on small screens.
+
+Wow! Ok, we have our work cut out for us. But wait til you see how easy this is! There'll be links to our actual
+production components involved (all open-source) at the end.
+
+## Pagination Schema
+
+Let's tackle the first part of this, which is: how do we adapt the
+[GraphQL connection spec](https://facebook.github.io/relay/graphql/connections.htm) in order to hold necessary
+information that a UI might need? Generally we want the UI's to be as simple as possible, and so if the server
+could construct a suitable pagination schema, that would be preferable. The simpler our UI, and the more business
+logic and good abstractions made in our GraphQL server, the more portable and reusable this all becomes.
+
+What kind of data does the UI need, in order to render a particular page of contents? Well, for a particular page
+we'd need to render the actual number it corresponds to. We'll need to know if this is the current page or not (so
+we can distinguish it in the UI from neighboring pages). And, we'll need to know the actual cursor (think:
+[opaque string](https://relay.dev/graphql/connections.htm)) that corresponds to this page number. It seems likely
+we'll need some sort of way to construct cursors from page numbers, on the server.
+
+So, check this out:
+
+```js
+// SDL
+type PageCursor {
+  cursor: String!
+  pageNumber: Int!
+  isCurrent: Boolean!
+}
+
+type PageCursors {
+  first: PageCursor
+  around: [PageCursor!]!
+  last: PageCursor
+  previous: PageCursor
+}
+```
+
+This is our pagination schema. Including a field of type `pageCursors` as a connection-level field, onto a
+connection, is sufficient for a UI to incredibly simply 'just render' a correct pagination bar always, and be able
+to hook up proper interactions. We can fully construct a simple UI (using Relay, shown in the next section) that
+can present and allow for the interactions desired, for windowed pagination.
+
+But, of course we're glossing over the implementation for such a `pageCursors` type, so let's check that out before
+looking at how a client might consume this.
+
+Our backing API's largely still paginate via offsets, and not cursors. That is, they accept page/size or
+size/offset style arguments. We use [graphql-relay-js](https://github.com/graphql/graphql-relay-js), which includes
+helpers to make sure types and resolvers are compatible with some Relay expectations. So, we use this library to
+generate our cursors, and can convert the cursor to an offset. A page of 4 with a size of 10, returns the elements
+numbered 30 - 39 in that list. So a page of 4 (and size of 10), is equivalent to an offset of 29 (and size of 10).
```

Comment:
Is it true that offset is 0-based this way? I'd expect the "first" page to have offset of 0, so the 4th to have offset of 30.